### DISCUSSION TITLE
use scaling and offset props by default

### DISCUSSION THREAD

**Discussion by iferencik** [developmentseed/titiler#803] [General]
I am building a STAC catalog in Azure Blob. The data are nightly mosaics for [nighttime lights](https://eogdata.mines.edu/products/vnl/#daily) from Colorado School of Mines and are converted to COGs before being inserted into the catalog. The original files are GeoTiffs rounded to 100th place and in average every COG occupies approx. 4 to 5 GB on disk. It occured to me that with some preprocessing (setting negative radiance to nodata, thresholding max radiance and using  a scale factor ) the data type can be changed to 2 byte unsigned integer halving thus the output COG size (2GB and less).

titiler can display the COGs but ignores the scale factor and shows the data to be of int type not float as opposed to  QGIS/GDAL. I think this might be a tricky one but in essence it is relatively easy to use the scale factor in the reader and I was wondering if there is a good/specific  reason why this is not happening.

The way I imagine this is through a config setting, either inside the server or as an env variable, or both, something like UNSCALE=TRUE/FALSE

I am looking forward to hear your thoughts on this

**Comment by vincentsarago** [comment]
TiTiler endpoints have an `unscale` option 

https://developmentseed.org/titiler/endpoints/cog/#tiles

> unscale (bool): Apply dataset internal Scale/Offset.

**Reply by iferencik** [reply]
@vincentsarago thanks for pointing that out.
how about adding this param to the info endpoint or including it in the response? I feel the second would be easier.
i think this is something useful, at least i find it useful. If i could learn a dataset has a scale factor i would be able to include the unscale param in the query.

**Reply by iferencik** [reply]
@vincentsarago 

i realized the info response can include scale and offset. But i can't see it here

https://titiler.undpgeohub.org/cog/info?url=https://undpgeohub.blob.core.windows.net/userdata/9426cffc00b069908b2868935d1f3e90/datasets/SVDNB_npp_d20240125_20240322094925.rade9di_ss.tif/SVDNB_npp_d20240125_20240322094925.rade9di_ss_band1.tif?c3Y9MjAyMy0xMS0wMyZzcz1iJnNydD1vJnNlPTIwMjUtMDMtMjJUMTAlM0E1NCUzQTI1WiZzcD1yJnNpZz1QQUoxbk5KRW5Xa010b1hkJTJCSzZYWng0SmkyZjJWa25hYSUyQlA3bFRtREhFTSUzRA==

but the point endpoint is using it
https://titiler.undpgeohub.org/cog/point/3.659851278742252,0.2162305742944568?url=https%3A%2F%2Fundpgeohub.blob.core.windows.net%2Fuserdata%2F9426cffc00b069908b2868935d1f3e90%2Fdatasets%2FSVDNB_npp_d20240125_20240322094925.rade9di_ss.tif%2FSVDNB_npp_d20240125_20240322094925.rade9di_ss_band1.tif%3Fsv%3D2023-11-03%26ss%3Db%26srt%3Do%26se%3D2025-03-22T10%253A51%253A38Z%26sp%3Dr%26sig%3DxBrOjZZ%252BLv9olgXhuhNh5FDClAB4Wb%252BWXeRcu7GOVcw%253D&bidx=1&unscale=True

https://titiler.undpgeohub.org/cog/point/3.659851278742252,0.2162305742944568?url=https%3A%2F%2Fundpgeohub.blob.core.windows.net%2Fuserdata%2F9426cffc00b069908b2868935d1f3e90%2Fdatasets%2FSVDNB_npp_d20240125_20240322094925.rade9di_ss.tif%2FSVDNB_npp_d20240125_20240322094925.rade9di_ss_band1.tif%3Fsv%3D2023-11-03%26ss%3Db%26srt%3Do%26se%3D2025-03-22T10%253A51%253A38Z%26sp%3Dr%26sig%3DxBrOjZZ%252BLv9olgXhuhNh5FDClAB4Wb%252BWXeRcu7GOVcw%253D&bidx=1


it this a bug?

**Reply by vincentsarago** [reply]
I'm not sure to get this, the `scale/offset` values should already be in the `Info` response https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/models.py#L80-L81

**Reply by iferencik** [reply]
gdalinfo shows it as well

gdalinfo "/vsicurl/https://undpgeohub.blob.core.windows.net/userdata/9426cffc00b069908b2868935d1f3e90/datasets/SVDNB_npp_d20240125_20240322094925.rade9di_ss.tif/SVDNB_npp_d20240125_20240322094925.rade9di_ss_band1.tif?sv=2023-11-03&ss=b&srt=o&se=2025-03-22T10%3A54%3A25Z&sp=r&sig=PAJ1nNJEnWkMtoXd%2BK6XZx4Ji2f2Vknaa%2BP7lTmDHEM%3D"



it looks like there is an issue in the info endpoint

**Reply by vincentsarago** [reply]
🤦 that's a bug https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/io/rasterio.py#L269-L273

**Reply by iferencik** [reply]
yes, indeed. But an easy one :)

**Reply by iferencik** [reply]
setting offset to 0 would not work either.

**Reply by vincentsarago** [reply]
https://github.com/cogeotiff/rio-tiler/pull/687 should fix this

**Reply by vincentsarago** [reply]
nice catch 🙏

**Reply by iferencik** [reply]
@vincentsarago do you know if the scale is available inside the algo call method?

https://github.com/UNDP-Data/geo-cogserver/blob/main/src/cogserver/algorithms/rca.py#L52


concretely, is there a way to get this info from ImageData object?

**Reply by vincentsarago** [reply]
> concretely, is there a way to get this info from ImageData object?

Sadly we do not forward scale/offset to the ImageData. This could be added in https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/reader.py#L272-L279

Feel free to start an issue/PR in rio-tiler 🙏

**Reply by iferencik** [reply]
@vincentsarago , i just check the same source and it was cleat the meta are tags.

rasterio reader has  the profile attr and I feel this should contain the scale and offset in case they are set.



 I would be in favor of adding  the profile attr to the ImageData


```


 return ImageData(
            data,
            bounds=out_bounds,
            crs=dataset.crs,
            band_names=[f"b{idx}" for idx in indexes],
            dataset_statistics=dataset_statistics,
            metadata=dataset.tags(),
            profile=dataset.profile
        )
```

in this way the algo developer has access to several useful props (nodata, etc)
the use  case here is that algos can do sophisticated things, in our case we are identifying the pixels that have chnaged between two sat images and are not clouds, so mask, nodata, and other params are useful.

**Reply by vincentsarago** [reply]
the problem is that `ImageData` are not just of single image, it might make sense for the output of rasterio Reader but for STAC or Mosaic it might be a bit more tricky

as I said, feel free to open an issue on rio-tiler and we can discuss more there
